go/nbs/manifest_cache.go

```diff
@@ -27,18 +31,49 @@ type manifestCache struct {
 	maxSize   uint64
 	lru       list.List
 	cache     map[string]manifestCacheEntry
+	locked    map[string]struct{}
+	cond      *sync.Cond
 }
 
 // Get() checks the searches the cache for an entry. If it exists, it moves it's
 // lru entry to the back of the queue and returns (value, true). Otherwise, it
 // returns (nil, false).
 func (mc *manifestCache) Get(db string) (contents manifestContents, present bool) {
+	mc.cond.L.Lock()
+	defer mc.cond.L.Unlock()
+
 	if entry, ok := mc.entry(db); ok {
 		contents, present = entry.contents, true
 	}
 	return
 }
 
+func (mc *manifestCache) Lock(db string) {
+	mc.cond.L.Lock()
+	defer mc.cond.L.Unlock()
+
+	for {
+		_, ok := mc.locked[db]
+		if ok {
+			mc.cond.Wait()
+		} else {
+			mc.locked[db] = struct{}{}
+			break
+		}
+	}
+}
+
+func (mc *manifestCache) Unlock(db string) {
+	mc.cond.L.Lock()
+	defer mc.cond.L.Unlock()
+
+	_, ok := mc.locked[db]
+	d.PanicIfFalse(ok)
+	delete(mc.locked, db)
+
+	mc.cond.Broadcast()
```

Comment:
It's unfortunate that we wake up everyone who's waiting on this every time when we know that only goroutines that were waiting on the lock for |db| actually have a chance of making progress :-/